File: TCE_IT_5S/Information Security Lab/Experiment 3/rsa.py
def Encrypt(p,pub):
    alpha="abcdefghijklmnopqrstuvwxyz"
    l=[]
    cipher=[]
    for i in p:
        l.append(alpha.index(i))
    for i in l:
        c=pow(i,pub[0],pub[1])
        cipher.append(c)
    return cipher
def Decrypt(c,pvt):
    l=[]
    alpha="abcdefghijklmnopqrstuvwxyz"
    for i in c:
        x=pow(i,pvt[0],pvt[1])
        l.append(alpha[x])
    return "".join(l)
def mul_inv(a,b):
    a,b=max(a,b),min(a,b)
    x=a
    t1=0
    t2=1
    while(b!=0):
        q=a//b
        r=a%b
        a=b
        b=r
        t=t1-q*t2
        t1=t2
        t2=t
    if a==1:
        return t1%x
    else:
        return "No inverse"
    
while(True):
    print("1.Key Gen\n2.Encrypt\n3.Decrypt")
    c=int(input("enter choice"))
    if c==1:
        pub,pvt=[],[]
        p=int(input("Enter prime:"))
        q=int(input("Enter prime:"))
        n=p*q
        phy=(p-1)*(q-1)
        e=int(input("Enter Public Key"))
        d=mul_inv(phy,e)
        pub.append(e)
        pvt.append(d)
        pub.append(n)
        pvt.append(n)
        print("public Key",pub)
        print("private key",pvt)

    if c==2:
        pt=input("Enter Text")
        pub=list(map(int,input("Enter pub key:").split(",")))
        print(Encrypt(pt,pub))
    if c==3:
        cipher=list(map(int,input("Enter cipher").split(",")))
        pub=list(map(int,input("Enter pub key:").split(",")))
        print(Decrypt(cipher,pvt))
        
        

File: TCE_IT_5S/Information Security Lab/Experiment 3/RSA.java
import java.util.*;
import java.math.*;

public class RSA {

	public boolean gcd(BigInteger e, BigInteger phi) {
		BigInteger gcd_value = e.gcd(phi);
		if(gcd_value.compareTo(BigInteger.valueOf(1))==0) {
			return true;
		}
		else {
			return false;
		}
	}
	
	public BigInteger decryptkey(BigInteger phi, BigInteger e) {
		BigInteger a=phi,b=e,q,r,t,t1,t2;
		t1=BigInteger.valueOf(0);
		t2=BigInteger.valueOf(1);
		while(!(b.equals(BigInteger.valueOf(0)))) {
			q=a.divide(b);
			r=a.mod(b);
			t=t1.subtract(q.multiply(t2));
			a=b;
			b=r;
			t1=t2;
			t2=t;
		}
		if(t1.compareTo(BigInteger.valueOf(0))==-1) {
			t1=t1.add(phi);
		}
		return t1;
	}
	
	public BigInteger encrypt(BigInteger plain, BigInteger e, BigInteger n) {
		return plain.modPow(e, n);
	}
	
	public BigInteger decrypt(BigInteger cipher, BigInteger d, BigInteger n) { {
		return cipher.modPow(d, n);
	}
	
	public static void main(String[] args) {
		Scanner Sc = new Scanner(System.in);
		RSA rsa = new RSA();
        String  alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

		//-----------------------------------------------------------------------------------	
		System.out.println("\n~~Key Generatiion\n");
        System.out.print("Enter the prime number value of p : ");
        BigInteger p = Sc.nextBigInteger();
        System.out.print("Enter the prime number value of q : ");
        BigInteger q = Sc.nextBigInteger();
        n = p.multiply(q);
        BigInteger one = BigInteger.valueOf(1);
                    
        BigInteger phi = (p.subtract(one)).multiply((q.subtract(one)));

        while(true) {
        	System.out.print("Enter the value of e : ");
            e = Sc.nextBigInteger();
            if(rsa.gcd(e,phi)==true) {
            	break;
            }
            else {
            	System.out.println("\n!!! e value is not accepted, Enter again");
            }
        }
        d = rsa.decryptkey(phi, e);
        System.out.println();
        System.out.println("Public Key --> ("+e+", "+n+")");
        System.out.println("Private Key --> ("+d+", "+n+")");
		//---------------------------------------------------------------------------------------
		
		//---------------------------------------------------------------------------------------
		System.out.print("Enter value of n : ");
        n=Sc.nextBigInteger();
        System.out.print("Enter encrypt key : ");
        e=Sc.nextBigInteger();
        System.out.print("Enter the character : ");
        String pt = Sc.next().toUpperCase();
        System.out.print("Encrypted Keys : ");
        for(int i=0;i<pt.length();i++) {
        	BigInteger plain = BigInteger.valueOf(alpha.indexOf(pt.charAt(i)));
            System.out.print(rsa.encrypt(plain, e, n)+" ");
        }
        //-----------------------------------------------------------------------------------------
        
        //-----------------------------------------------------------------------------------------
		System.out.print("Enter value of n : ");
        n=Sc.nextBigInteger();
        System.out.print("Enter decrypt key : ");
        d=Sc.nextBigInteger();
        System.out.print("Enter number of characters : ");
        int num_char=Sc.nextInt();
        BigInteger[] Cipher_arr=new BigInteger[num_char];
        for(int i=0;i<num_char;i++) {
        	Cipher_arr[i]=Sc.nextBigInteger();
        }
        System.out.print("Plain text:");
        for(int i=0;i<num_char;i++) {
        	System.out.print(alpha.charAt(rsa.decrypt(Cipher_arr[i], d, n).intValue()));
        }
		//-------------------------------------------------------------------------------------------
	
	}

}
